dm/master/openapi.go

```diff
@@ -133,17 +134,100 @@ func (s *Server) DMAPIDeleteSource(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.StartRelayRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	// update relay related in source cfg
+	sourceCfg.EnableRelay = true
+	if req.RelayBinlogName != nil && sourceCfg.RelayBinLogName != *req.RelayBinlogName {
+		sourceCfg.RelayBinLogName = *req.RelayBinlogName
+	}
+	if req.RelayBinlogGtid != nil && sourceCfg.RelayBinlogGTID != *req.RelayBinlogGtid {
+		sourceCfg.RelayBinlogGTID = *req.RelayBinlogGtid
+	}
+	if req.RelayDir != nil && sourceCfg.RelayDir != *req.RelayDir {
+		sourceCfg.RelayDir = *req.RelayDir
+	}
+	if purge := req.Purge; purge != nil {
+		if purge.Expires != nil && sourceCfg.Purge.Expires != *purge.Expires {
+			sourceCfg.Purge.Expires = *purge.Expires
+		}
+		if purge.Interval != nil && sourceCfg.Purge.Interval != *purge.Interval {
+			sourceCfg.Purge.Interval = *purge.Interval
+		}
+		if purge.RemainSpace != nil && sourceCfg.Purge.RemainSpace != *purge.RemainSpace {
+			sourceCfg.Purge.RemainSpace = *purge.RemainSpace
+		}
+	}
+	// update current source relay config before start relay
+	if err := s.scheduler.UpdateSourceCfg(sourceCfg); err != nil {
+		return err
+	}
+	return s.scheduler.StartRelay(sourceName, []string{req.WorkerName})
 }
 
 // DMAPIStopRelay url is:(DELETE /api/v1/sources/{source-id}/relay).
 func (s *Server) DMAPIStopRelay(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.WorkerNameRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	return s.scheduler.StopRelay(sourceName, []string{req.WorkerName})
 }
 
 // DMAPIGetSourceStatus url is:(GET /api/v1/sources/{source-id}/status).
 func (s *Server) DMAPIGetSourceStatus(ctx echo.Context, sourceName string) error {
-	return nil
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	workerName := ""
+	if w := s.scheduler.GetWorkerBySource(sourceName); w != nil {
+		workerName = w.BaseInfo().Name
+	}
+	resp := openapi.SourceStatus{
+		EnableRelay: sourceCfg.EnableRelay,
+		SourceName:  sourceCfg.SourceID,
+		WorkerName:  workerName,
+	}
+	// only query worker if source is bound to a worker
+	queryWorker := workerName != ""
+	if queryWorker {
+		var statusResp *pb.QueryStatusResponse
+		workerResp := s.getStatusFromWorkers(ctx.Request().Context(), []string{sourceName}, "", sourceCfg.EnableRelay)
+		for idx := range workerResp {
+			if workerResp[idx].SourceStatus.Worker == workerName {
+				statusResp = workerResp[idx]
+				break
+			}
+		}
+		if statusResp == nil {
+			// No response from worker and master means that the current query source has not been created.
+			// and this should not happen here.
+			return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
```

Comment:
please check if this will happen, if so, don't use ErrSchedulerSourceCfgNotExist